google/cloud/spanner/numeric.h

```diff
@@ -0,0 +1,269 @@
+    return os << n.ToString();
+  }
+
+ private:
+  friend StatusOr<Numeric> internal::MakeNumeric(std::string s);
+  explicit Numeric(std::string rep) : rep_(std::move(rep)) {}
+  std::string rep_;  // a valid and canonical NUMERIC representation
+};
+
+namespace internal {
+
+// Like `std::to_string`, but also supports Abseil 128-bit integers.
+template <typename T>
+std::string ToString(T value) {
+  return std::to_string(value);
+}
+std::string ToString(absl::int128 value);
+std::string ToString(absl::uint128 value);
+
+// Forward declarations.
+Status DataLoss(std::string message);
+StatusOr<Numeric> MakeNumeric(std::string s, int exponent);
+
+}  // namespace internal
+
+/**
+ * Construction from a string, in decimal fixed- or floating-point formats
+ * (where there must be digits either before or after the decimal point):
+ *
+ *    [-+]?[0-9]+(.[0-9]*)?([eE][-+]?[0-9]+)?
+ *    [-+]?.[0-9]+([eE][-+]?[0-9]+)?
+ *
+ * Fails on syntax errors or if the conversion would yield a value outside
+ * the NUMERIC range. If the argument has more than `kFracPrec` digits after
+ * the decimal point it will be rounded, with halfway cases rounding away
+ * from zero.
+ */
+StatusOr<Numeric> MakeNumeric(std::string s);
+
+/**
+ * Construction from a double. Fails on NaN or any argument outside the
+ * NUMERIC value range (which includes infinities). If the argument has more
+ * than `kFracPrec` digits after the decimal point it will be rounded, with
+ * halfway cases rounding away from zero.
+ */
+StatusOr<Numeric> MakeNumeric(double d);
+
+/**
+ * Construction from an integer `i`, scaled by 10^`exponent`. Fails on any
+ * (scaled) argument outside the NUMERIC value range.
+ */
+template <typename T, typename std::enable_if<
+                          std::numeric_limits<T>::is_integer, int>::type = 0>
+StatusOr<Numeric> MakeNumeric(T i, int exponent = 0) {
+  return internal::MakeNumeric(internal::ToString(i), exponent);
+}
+
+/**
+ * Conversion to the closest double value, with possible loss of precision.
+ * Always succeeds (i.e., can never overflow, assuming a double can hold
+ * values up to 10^(kIntPrec+1)).
+ */
+inline double ToDouble(Numeric const& n) {
+  return std::atof(n.ToString().c_str());
+}
+
+/**
+ * Conversion to the nearest integer value, scaled by 10^`exponent`, rounding
+ * halfway cases away from zero. Fails when the destination type cannot hold
+ * that value.
+ *
+ * @par Example
+ *
+ * @code
+ * spanner::Numeric n = spanner::MakeNumeric(123456789, -2).value();
+ * assert(n.ToString() == "1234567.89");
+ * assert(spanner::ToInteger<int>(n).value() == 1234568);
+ * assert(spanner::ToInteger<int>(n, 2).value() == 123456789);
+ * @endcode
+ */
+///@{
+template <typename T,
+          typename std::enable_if<std::numeric_limits<T>::is_integer &&
+                                      !std::numeric_limits<T>::is_signed,
+                                  int>::type = 0>
+StatusOr<T> ToInteger(Numeric const& n, int exponent = 0) {
```

Comment:
This is a somewhat large function template at 30 lines, and I'm thinking about possible binary bloat from it being stamped out into the binary for each `T` that this template is instantiated with. If that seems like a legit concern, we could have the majority of the body of this function in a single non-templated `StatusOr<absl::int128> ToIntegerImpl(...)` that returns the value as an `absl::int128`, then the templated wrapper only needs to look at the returned int128 and see if it fits in the requested `T`. Of course, this solution assumes that `int128` is big enough to hold any `T` that the caller provides. I think that'd be a fine assumption.

We'd also need to do the same thing for `absl::uint128`, but it'd follow the same pattern for the same reasons.

WDYT? Do we care about potential binary bloat here?